nasa/ow_simulator
# Oceanwater 614 implement a force torque sensor for the lander arm

## Linked Issues:
| EPIC ⚡| [OCEANWATER-614 / Implement a force torque sensor for the lander arm ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-614) |
| :----------- | :----------- |
| Jira Ticket 🎟️   | ~[OCEANWATER-615 / Introduce a force torque sensor for the dist_patch joint ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-615)~ |
| Jira Ticket 🎟️   | ~[OCEANWATER-616 / Implement guarded move using the ft sensor ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-616)~ |
| Jira Ticket 🎟️   | ~[OCEANWATER-617 / Introduce sufficient noise to model a realistic sensor ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-617)~ |
| Jira Ticket 🎟️   | ~[OCEANWATER-618 / Model potential faults or failure for the ft sensor ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-618)~ |
| Jira Ticket 🎟️   | ~[OCEANWATER-619 / Update the user guide to include utilization of the FT Sensor ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-619)~ |
| Jira Ticket 🎟️   | [OCEANWATER-620 / Document sensor response when loading a sample with varying mass into scoop ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-620) [DESCOPED] |
| Jira Ticket 🎟️   | [OCEANWATER-646 / Model FT sensor drift ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-646) [DESCOPED] |
| Jira Ticket 🎟️   | [OCEANWATER-647 / Increased noise of ft sensor in correlation with certain events ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-647) [DESCOPED] |
| Github :octocat:  | #94, #96, #106, #109, #110 |

## Summary of Changes
* Introduced a force torque sensor for the **dist_patch** joint (highlighted below in red)
![image](https://user-images.githubusercontent.com/606033/113636813-19c2ff00-9628-11eb-854f-e7611018467f.png)
* Implement guarded move using readings of the ft sensor
  - Removed the old method
* Set FT sensor update rate to 200 Hz
  - Note that the [Barret FT sensor](https://web.barrett.com/files/Low-Profile_Force-Torque_Sensor_FT-02-2011.pdf) runs at 250 Hz by default and has a maximum update rate of 2500 Hz. However, the **gazebo_ros_ft_sensor** is only able to provide a maximum update rate of 200 Hz. 
* Set sensor Gaussian noise value to 0.2
  - Noise value is based on generated noise profile of the [Barret FT sensor](https://web.barrett.com/files/Low-Profile_Force-Torque_Sensor_FT-02-2011.pdf)

| Barret FT Sensor Noise | Simulated FT Sensor Noise |
| :----------: | :----------: |
|![image](https://user-images.githubusercontent.com/606033/112533125-daaec700-8d66-11eb-9eea-54a516a4fa5b.png)|![image](https://user-images.githubusercontent.com/606033/112551442-3df82380-8d7e-11eb-9854-24affb53d0bc.png)|
|![image](https://user-images.githubusercontent.com/606033/112531875-59a30000-8d65-11eb-9b7b-73c117321c59.png)|![image](https://user-images.githubusercontent.com/606033/112532039-9242d980-8d65-11eb-8923-e6b8cd18286c.png)|
* Modeled three types of faults for the ft_sensor
  - zero signal
  - signal bias
  - increase signal noise

| Fault | Signal |
| :----------- | :----------- |
| Zero Signal |![image](https://user-images.githubusercontent.com/606033/113361141-22ea5e00-9300-11eb-90d0-e47af8027905.png)|
| Signal Bias |![image](https://user-images.githubusercontent.com/606033/113361160-2d0c5c80-9300-11eb-898c-3d20cc192160.png)|
| Increased Signal Noise |![image](https://user-images.githubusercontent.com/606033/113361180-3695c480-9300-11eb-98cc-207035526dde.png)|

## Test
 1) Launch the simulation
```bash
roslaunch ow (atacma_y1a.launch | europa_terminator_workspace.launch)
```
2) While the simulation running open another terminal and use rostopic to listen to the **ft_sensor_dist_pitch** topic
```bash
rostopic echo /ft_sensor_dist_pitch
```
* You should see a stream of **geometry_msgs/WrenchStamped** messages with relatively stable values for force and torque.
3) Execute any arm activity and you should see the stream of values fluctuate.

4) While the simulation running invoke guarded move with varying set of valid arguments in the same session.
```bash
rosservice call /arm/guarded_move -- True 0 0 0 0 0 0 0
rosservice call /arm/guarded_move -- False 1.7 0.3 0.5 0 0 1 0.7
rosservice call /arm/guarded_move -- False 1.7 -0.3 0.5 0 0 1 0.7
rosservice call /arm/guarded_move -- False 1.9 -0.4 0.5 0 0 1 0.7
rosservice call /arm/guarded_move -- False 1.9 0.5 0.5 0 0 1 0.7
rosservice call /arm/guarded_move -- False 1.7 0.0 0.5 1 1 1 0.7
rosservice call /arm/guarded_move -- False 1.7 0.0 0.5 0.5 -0.5 1 0.8
```
The ground is detected with no delay or premature detection.

5) Use the **dyanmic_reconfigure** UI to apply any of the three faults (zero signal, signal bias, increase signal noise) and see the signal changes as shown the faults table mentioned earlier
![image](https://user-images.githubusercontent.com/606033/113361344-a906a480-9300-11eb-9f63-6e82fa5c4ac5.png)
>Note: you can apply them individually or in combination
